design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,506 @@
+Arm architecture (e.g. for testing), the most critical use is when transferred
+to an FPGA (i.e. not an Arm architecture).
+
+That said, if the operation that this FPGA performs becomes popular there may be
+a need to run the code directly on CPU's in the future.
+
+The requirements on Arm ABI's from this use-case are relatively small since the
+main focus is around running on an FPGA.  We believe it adds weight to both the
+need for performance and familiarity of programmers.  This belief comes from the
+estimate that this may lead to bit-precise types being used in performance
+critical code in the future, and that it may mean that bit-precise types are
+used on Arm architectures when testing FPGA descriptions (where ease of
+debugging can be prioritized).
+
+
+24-bit Color
+~~~~~~~~~~~~~
+
+Some image file-types use 24-bit color.  The new ``_BitInt`` type may be used to
+hold such information.
+
+As it stands we do not know of any specific reason to use a bit-precise integral
+type as opposed to a structure of three bytes for these data types.
+
+If used for 24-bit color we believe that the performance of standard arithmetic
+operations would not be critical.  This because each 24-bit pixel usually
+represents three 8-bit colors so operations would unlikely be performed on the
+single value as a whole.
+
+We also believe that if used for 24-bit color it would be helpful to specify a
+size and alignment scheme such that an array of ``_BitInt(24)`` is well packed.
+
+
+Networking Protocols
+~~~~~~~~~~~~~~~~~~~~
+
+Many networking protocols have packed structures in order to minimize data sent
+over the wire.  In order to be perfectly packed the code will need to use
+bit-fields rather than bit-precise types for storage, since the bit-precise types
+must be accessible and hence at least byte-aligned.
+
+The incentives to use bit-precise integral types for networking code would be in
+order to maintain the best representation of the operation that is being
+performed.
+
+One negative of using bit-precise integral types for networking code would be
+that idioms like ``if (x + y > max_representable)`` where ``x`` and ``y`` have
+been loaded from small bit-fields would no longer be viable.  We have seen such
+idioms for small values in networking code in the Linux kernel.  These are
+intuitive to write but if ``x`` and ``y`` were to bit-precise types would not
+work as expected.
+
+If used in code handling networking protocols, our estimate is that the
+arithmetic manipulation performed on such values will not be the main
+performance bottleneck in networking protocols.  This estimation comes from the
+belief that networking is often IO bound, and that small packed values in
+networking protocols tend to have limited arithmetic performed on them.
+
+Hence we believe that ease of debugging of values in registers may be more
+critical than performance concerns in this use-case.
+
+
+To help the compiler optimize (e.g. for auto vectorization)
```

Comment:
```suggestion
To help the compiler optimize (for example, for auto vectorization)
```

Reply:
Adjusted this to `(possibly for auto vectorization)` -- hope that isn't falling into another ease-of-parsing pitfall ;-)